app-marketplace-v2-spec/architecture.md

```diff
@@ -0,0 +1,146 @@
+3. Functions may accept both structured JSON inputs, as well as "raw" command strings with options.
+4. Dialogs and commands that are fully valid against SlashCommands definition are expected to submit structured data.
+5. Widgets "hardcode" their invocation of functions, so they can do as they please.
+6. Slash Commands invoke the closest-matched function. If the command string passes validation, it is submitted as JSON, otherwise as raw text. Raw text is always submitted.
+7. Expandable context provides functions with most data dependencies pre-fetched, as instructed by the App Manifest, or the function binding.
+
+### Interactive Inputs.
+See [Interactive Inputs](commands.md) for more.
+  - **Slash Commands** with autocomplete, execute functions.
+  - **Widgets** that can be bound to directly execute functions (Post menu items, channel menu, interactive post buttons, etc). In the future they may be extended to become in-place forms or autocomplete boxes.
+  - **Modal Dialogs** and **Bot Dialogs**, that gather/edit parameter values, then execute a function.
+
+#### Slash Commands
+1. Slash commands will have interactive auto-complete, available in the usual message box in Mattermost. All declared commands that are not "hidden" will be available for autocomplete.
+2. #TODO we need to be able to re-define what is visible (valid?) based on the state already entered, and hide the irrelevant options. @iomodo can you take a look at this?
+3. #STRETCH a "slash command" widget that would execute a pre-configured autocomplete command, in the given context (team/channel or Bot DM), and display the result.
+4. Slash commands may have a --modal option that would launch a modal to gather the inputs, rather than autocomplete. The parameters in the slash command are used as the default values in the modal.
+5. All valid slash commands must be declared in the App manifest
+
+#### Widgets
+1. A widget is a UX element "hard-wired" to execute a function upon a simple user interaction.
+2. Widgets must be pre-declared in the manifest, or inserted in Posts.
+3. Widgets' visibility in Mattermost UX can be toggled contextually (user, channel, team prefs) #TODO how?
+4. Widgets' visibility in Mattermost UX can be toggled dynamically, via a lookup API #TODO how?
+5. Widgets can be stateful (e.g. on/off button)
+
+#### Dialogs
+- #TODO
+
+### Change notifications.
+1. Subjects
+  - *App Lifecycle* Installed/Uninstalled
+  - MessagePosted (matches the existing webhook)
+      Scoped to: User, Channel.
+  - ChannelCreated
+  - UserCreated
+  - UserJoinedChannel
+  - UserJoinedTeam
+  - UserLeftChannel
+  - UserLeftTeam
+  - UserUpdated
+2. Expandable context provides functions with most data dependencies pre-fetched, as instructed by the App Manifest, or the function binding.
+3. Overall scope in the App Manifest. #STRETCH individual scoped subscriptions for performance - channel, user, etc.
+
+### Autocomplete lookups
+- #TODO
+
+### Lifecycle - Install
```

Comment:
Can we say more about the lifecycle?
I assume there will be apps controlled by us, installed in the cluster, etc.
And there will be third-party apps, which will be running somewhere else. Correct?